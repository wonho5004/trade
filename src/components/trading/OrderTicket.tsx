'use client';

import { useEffect, useState } from 'react';
import type { ChangeEvent, FormEvent } from 'react';

import { useChartStore } from '@/stores/chartStore';

const defaultForm = {
  symbol: 'BTCUSDT',
  side: 'buy',
  quantity: '0.01',
  price: ''
};

export function OrderTicket() {
  const [form, setForm] = useState(defaultForm);
  const currentSymbol = useChartStore((state) => state.symbol);

  useEffect(() => {
    setForm((prev) => ({ ...prev, symbol: currentSymbol }));
  }, [currentSymbol]);

  const handleChange = (event: ChangeEvent<HTMLInputElement | HTMLSelectElement>) => {
    const { name, value } = event.target;
    setForm((prev) => ({ ...prev, [name]: value }));
  };

  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    // 주문 제출 로직은 CCXT 주문 생성 모듈 연동 후 구현 예정입니다.
    console.info('주문 요청', form);
  };

  return (
    <form
      onSubmit={handleSubmit}
      className="flex w-full flex-col gap-4 rounded-lg border border-zinc-800 bg-zinc-950 p-6"
    >
      <div className="grid gap-2">
        <label className="text-sm text-zinc-400" htmlFor="symbol">
          거래쌍
        </label>
        <input
          id="symbol"
          name="symbol"
          value={form.symbol}
          onChange={handleChange}
          className="rounded border border-zinc-700 bg-zinc-900 px-3 py-2 text-sm text-zinc-100"
        />
      </div>
      <div className="grid gap-2">
        <label className="text-sm text-zinc-400" htmlFor="side">
          매매 방향
        </label>
        <select
          id="side"
          name="side"
          value={form.side}
          onChange={handleChange}
          className="rounded border border-zinc-700 bg-zinc-900 px-3 py-2 text-sm text-zinc-100"
        >
          <option value="buy">매수</option>
          <option value="sell">매도</option>
        </select>
      </div>
      <div className="grid gap-2">
        <label className="text-sm text-zinc-400" htmlFor="quantity">
          수량
        </label>
        <input
          id="quantity"
          name="quantity"
          value={form.quantity}
          onChange={handleChange}
          type="number"
          step="0.0001"
          min="0"
          className="rounded border border-zinc-700 bg-zinc-900 px-3 py-2 text-sm text-zinc-100"
        />
      </div>
      <div className="grid gap-2">
        <label className="text-sm text-zinc-400" htmlFor="price">
          지정가 (선택)
        </label>
        <input
          id="price"
          name="price"
          value={form.price}
          onChange={handleChange}
          type="number"
          step="0.1"
          min="0"
          className="rounded border border-zinc-700 bg-zinc-900 px-3 py-2 text-sm text-zinc-100"
        />
      </div>
      <button
        type="submit"
        className="rounded bg-emerald-500 px-4 py-2 text-sm font-semibold text-emerald-950 transition hover:bg-emerald-400"
      >
        주문 실행
      </button>
    </form>
  );
}
