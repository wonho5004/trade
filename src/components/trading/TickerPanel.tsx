'use client';

import { useEffect, useMemo } from 'react';
import { useShallow } from 'zustand/react/shallow';

import { useChartStore } from '@/stores/chartStore';
import { useMarketDataStore } from '@/stores/marketDataStore';

const formatNumber = (value: number | null, fractionDigits = 2) => {
  if (value === null || Number.isNaN(value)) {
    return '-';
  }
  return value.toLocaleString(undefined, {
    minimumFractionDigits: fractionDigits,
    maximumFractionDigits: fractionDigits
  });
};

export function TickerPanel() {
  const symbol = useChartStore((state) => state.symbol);
  const { status, lastPrice, priceChangePercent, highPrice, lowPrice, volume } = useMarketDataStore(
    useShallow((state) => ({
      status: state.status,
      lastPrice: state.lastPrice,
      priceChangePercent: state.priceChangePercent,
      highPrice: state.highPrice,
      lowPrice: state.lowPrice,
      volume: state.volume
    }))
  );
  const attachTicker = useMarketDataStore((state) => state.attach);

  useEffect(() => {
    if (!symbol) {
      return;
    }
    const detach = attachTicker(symbol);
    return detach;
  }, [symbol, attachTicker]);

  const isPositive = useMemo(() => (priceChangePercent ?? 0) >= 0, [priceChangePercent]);

  return (
    <section className="flex h-full flex-col justify-between rounded-lg border border-zinc-800 bg-zinc-950 p-6">
      <header className="flex items-baseline justify-between">
        <div>
          <p className="text-sm text-zinc-400">실시간 가격</p>
          <h2 className="text-lg font-semibold text-zinc-100">{symbol}</h2>
        </div>
        <span
          className={`text-sm font-semibold ${
            isPositive ? 'text-emerald-400' : 'text-rose-400'
          }`}
        >
          {status === 'connected' && priceChangePercent !== null
            ? `${formatNumber(priceChangePercent, 2)}%`
            : status === 'connecting'
            ? '연결 중...'
            : status === 'error'
            ? '에러'
            : '-'}
        </span>
      </header>

      <div className="mt-6 grid grid-cols-2 gap-4 text-sm text-zinc-400">
        <div>
          <p className="text-xs uppercase tracking-wide text-zinc-500">마지막 체결가</p>
          <p className="mt-2 text-base font-semibold text-zinc-100">
            {status === 'connected' ? formatNumber(lastPrice, 2) : '-'}
          </p>
        </div>
        <div>
          <p className="text-xs uppercase tracking-wide text-zinc-500">24h 고가</p>
          <p className="mt-2 text-base font-semibold text-zinc-100">{formatNumber(highPrice, 2)}</p>
        </div>
        <div>
          <p className="text-xs uppercase tracking-wide text-zinc-500">24h 저가</p>
          <p className="mt-2 text-base font-semibold text-zinc-100">{formatNumber(lowPrice, 2)}</p>
        </div>
        <div>
          <p className="text-xs uppercase tracking-wide text-zinc-500">24h 거래량</p>
          <p className="mt-2 text-base font-semibold text-zinc-100">{formatNumber(volume, 0)}</p>
        </div>
      </div>
    </section>
  );
}
