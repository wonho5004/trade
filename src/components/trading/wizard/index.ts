/**
 * Setup Wizard 컴포넌트 export
 */

export { SetupWizard } from './SetupWizard';
export { WelcomeStep } from './WelcomeStep';
export { ExperienceStep } from './ExperienceStep';
export { RecommendationStep } from './RecommendationStep';
export { CompleteStep } from './CompleteStep';
