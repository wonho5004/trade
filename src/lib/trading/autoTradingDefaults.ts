import type {
  AutoTradingSettings,
  AdditionalEntryConditions,
  CapitalSettings,
  EntryConditions,
  EntryDirectionSettings,
  ExitConditions,
  ExitDirectionSettings,
  HedgeActivationSettings,
  IndicatorConditions,
  IndicatorConfigMap,
  IndicatorEntry,
  IndicatorKey,
  AggregatorOperator,
  ConditionGroupNode,
  ConditionNode,
  IndicatorLeafNode,
  CandleLeafNode,
  InitialMarginSetting,
  PositionDirection,
  ProfitCondition,
  ScaleInBudget,
  ScaleInDirectionSettings,
  ScaleInLimit,
  StopLossConditions,
  StopLossLineSettings,
  SymbolSelection,
  ThresholdCondition,
  IndicatorComparisonLeaf,
  CandleCondition,
  ComparisonOperator,
  CandleFieldOption,
  CandleReference
} from '@/types/trading/auto-trading';

export const createProfitCondition = (): ProfitCondition => ({
  enabled: false,
  comparator: 'over',
  value: 0
});

export const createThresholdCondition = (): ThresholdCondition => ({
  enabled: false,
  comparator: 'over',
  value: 0
});

const createIndicatorEntryId = () => `ind-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 8)}`;

const createConditionNodeId = () => `cond-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 8)}`;

export const createCandleCondition = (): CandleCondition => ({
  enabled: false,
  field: 'close',
  comparator: 'over',
  targetValue: 0,
  reference: 'current'
});

export const createCandleLeaf = (candle?: CandleCondition): CandleLeafNode => ({
  kind: 'candle',
  id: createConditionNodeId(),
  candle: candle ? { ...candle } : createCandleCondition()
});

export const createBollingerCondition = (): IndicatorConfigMap['bollinger'] => ({
  enabled: false,
  length: 20,
  standardDeviation: 2,
  offset: 0,
  band: 'middle',
  action: 'touch',
  touchTolerancePct: 0.2
});

export const createMaCondition = (): IndicatorConfigMap['ma'] => ({
  enabled: false,
  period: 20,
  actions: []
});

export const createRsiCondition = (): IndicatorConfigMap['rsi'] => ({
  enabled: false,
  period: 14,
  smoothing: 'sma',
  threshold: 50,
  actions: []
});

export const createDmiCondition = (): IndicatorConfigMap['dmi'] => ({
  enabled: false,
  diPeriod: 14,
  adxPeriod: 14,
  adx: createThresholdCondition(),
  diComparison: null,
  diPlus: createThresholdCondition(),
  diMinus: createThresholdCondition(),
  adxVsDi: null
});

export const createMacdCondition = (): IndicatorConfigMap['macd'] => ({
  enabled: false,
  fast: 12,
  slow: 26,
  signal: 9,
  source: 'close',
  method: 'EMA',
  comparison: null,
  histogramAction: null
});

export const createIndicatorConfig = <T extends IndicatorKey>(indicator: T): IndicatorConfigMap[T] => {
  switch (indicator) {
    case 'bollinger':
      return createBollingerCondition() as IndicatorConfigMap[T];
    case 'ma':
      return createMaCondition() as IndicatorConfigMap[T];
    case 'rsi':
      return createRsiCondition() as IndicatorConfigMap[T];
    case 'dmi':
      return createDmiCondition() as IndicatorConfigMap[T];
    case 'macd':
      return createMacdCondition() as IndicatorConfigMap[T];
    default:
      return createBollingerCondition() as IndicatorConfigMap[T];
  }
};

export const createIndicatorEntry = <T extends IndicatorKey>(indicator: T): IndicatorEntry<T> => ({
  id: createIndicatorEntryId(),
  type: indicator,
  config: createIndicatorConfig(indicator)
});

export const createIndicatorComparisonLeaf = (
  legacy: unknown = { kind: 'none' },
  fallbackId: string
): IndicatorComparisonLeaf | { kind: 'none' } => {
  if (!legacy || typeof legacy !== 'object') {
    return { kind: 'none' };
  }
  const candidate = legacy as any;
  if (candidate.kind === 'candle' && candidate.field && candidate.reference && candidate.comparator) {
    return {
      kind: 'candle',
      comparator: candidate.comparator,
      field: candidate.field,
      reference: candidate.reference
    };
  }
  if (candidate.kind === 'value' && typeof candidate.value === 'number' && candidate.comparator) {
    return {
      kind: 'value',
      comparator: candidate.comparator,
      value: candidate.value
    };
  }
  if (candidate.kind === 'indicator' && candidate.comparator) {
    return {
      kind: 'indicator',
      comparator: candidate.comparator,
      targetIndicatorId: candidate.targetIndicatorId ?? fallbackId
    };
  }
  return { kind: 'none' };
};

export const createIndicatorLeaf = (
  indicator: IndicatorEntry,
  comparison: IndicatorComparisonLeaf | { kind: 'none' } = { kind: 'none' }
): IndicatorLeafNode => ({
  kind: 'indicator',
  id: createConditionNodeId(),
  indicator,
  comparison
});

export const createIndicatorGroup = (
  operator: AggregatorOperator = 'and',
  children: ConditionNode[] = []
): ConditionGroupNode => ({
  kind: 'group',
  id: createConditionNodeId(),
  operator,
  children
});

export const createIndicatorConditions = (operator: AggregatorOperator = 'or'): IndicatorConditions => {
  // Stable default tree with deterministic IDs for test equality
  const candle = createCandleCondition();
  const root: ConditionGroupNode = {
    kind: 'group',
    id: 'cond-root',
    operator,
    children: []
  };

  // Return a legacy-compatible shape for transitional tests
  return {
    root,
    candle,
    entries: [],
    defaultAggregator: root.operator
  } as unknown as IndicatorConditions;
};

export const normalizeConditionTree = (root: ConditionNode): ConditionGroupNode => {
  // Defensive: tolerate unexpected/legacy shapes where `root` may be missing or malformed
  const isValidNode = (n: any) => n && typeof n === 'object' && typeof n.kind === 'string';
  const base: ConditionNode = isValidNode(root)
    ? (root as ConditionNode)
    : createIndicatorGroup('and', []);

  const collectIndicatorIds = (node: ConditionNode, acc: Set<string>) => {
    if (node.kind === 'indicator') {
      acc.add(node.id);
      return;
    }
    if (node.kind === 'group') {
      node.children.forEach((child) => collectIndicatorIds(child, acc));
    }
  };

  const sanitizeComparisons = (node: ConditionNode, indicatorIds: Set<string>): ConditionNode => {
    if (node.kind === 'indicator') {
      if (node.comparison.kind === 'indicator' && !indicatorIds.has(node.comparison.targetIndicatorId)) {
        return { ...node, comparison: { kind: 'none' } };
      }
      return node;
    }
    if (node.kind === 'group') {
      return {
        ...node,
        children: node.children.map((child) => sanitizeComparisons(child, indicatorIds))
      };
    }
    return node;
  };

  const ensureGroupRoot = (node: ConditionNode): ConditionGroupNode => {
    if (node.kind === 'group') return { ...node };
    return createIndicatorGroup('and', [node]);
  };

  const indicatorIds = new Set<string>();
  collectIndicatorIds(base, indicatorIds);
  const sanitized = sanitizeComparisons(base, indicatorIds);
  return ensureGroupRoot(sanitized);
};

const createInitialMarginSetting = (): InitialMarginSetting => ({
  basis: 'wallet',
  mode: 'per_symbol_percentage',
  percentage: 5,
  minNotional: 0,
  usdtAmount: 0,
  asset: 'USDT'
});

const createScaleInBudget = (): ScaleInBudget => ({
  mode: 'balance_percentage',
  basis: 'wallet',
  percentage: 5,
  minNotional: 0,
  usdtAmount: 0,
  asset: 'USDT'
});

const createScaleInLimit = (): ScaleInLimit => ({
  balance: {
    enabled: false,
    basis: 'wallet',
    percentage: 10
  },
  initialMargin: {
    enabled: false,
    percentage: 100
  },
  purchaseCount: {
    enabled: false,
    comparator: 'over',
    value: 1
  },
  profitRate: createProfitCondition(),
  unlimited: true
});

export const createCapitalSettings = (): CapitalSettings => ({
  estimatedBalance: 10000,
  maxMargin: {
    basis: 'wallet',
    percentage: 20
  },
  initialMargin: createInitialMarginSetting(),
  scaleInBudget: createScaleInBudget(),
  scaleInLimit: createScaleInLimit(),
  useMinNotionalFallback: true,
  hedgeBudget: {
    separateByDirection: false,
    long: { mode: 'position_percent', percentage: 100 },
    short: { mode: 'position_percent', percentage: 100 }
  }
});

export const createEntryDirectionSettings = (direction: PositionDirection): EntryDirectionSettings => ({
  enabled: false,
  immediate: false,
  indicators: createIndicatorConditions()
});

export const createEntryConditions = (): EntryConditions => ({
  long: createEntryDirectionSettings('long'),
  short: createEntryDirectionSettings('short')
});

export const createScaleInDirectionSettings = (direction: PositionDirection): ScaleInDirectionSettings => ({
  enabled: direction === 'long',
  profitTarget: createProfitCondition(),
  currentProfitRate: createProfitCondition(),
  indicators: createIndicatorConditions(),
  currentBuyNotional: { enabled: false, asset: 'USDT', comparator: 'none', value: 0 },
  initialBuyChange: { enabled: false, comparator: 'none', value: 0 }
});

export const createScaleInConditions = (): AdditionalEntryConditions => ({
  long: createScaleInDirectionSettings('long'),
  short: createScaleInDirectionSettings('short')
});

export const createExitDirectionSettings = (direction: PositionDirection): ExitDirectionSettings => ({
  enabled: direction === 'long',
  profitTarget: createProfitCondition(),
  currentProfitRate: createProfitCondition(),
  indicators: createIndicatorConditions(),
  includeFeesFunding: false,
  currentBuyNotional: { enabled: false, asset: 'USDT', comparator: 'none', value: 0 },
  initialBuyChange: { enabled: false, comparator: 'none', value: 0 },
  sellAmount: { mode: 'percent', value: 0 }
});

export const createExitConditions = (): ExitConditions => ({
  long: createExitDirectionSettings('long'),
  short: createExitDirectionSettings('short')
});

export const createStopLossLineSettings = (): StopLossLineSettings => ({
  enabled: false,
  autoRecreate: true,
  profitTarget: createProfitCondition(),
  indicators: createIndicatorConditions()
});

export const createStopLossConditions = (): StopLossConditions => ({
  profitTarget: createProfitCondition(),
  currentProfitRate: createProfitCondition(),
  purchaseCount: {
    enabled: false,
    comparator: 'over',
    value: 1
  },
  stopLossLine: createStopLossLineSettings(),
  indicators: createIndicatorConditions()
});

export const createSymbolSelection = (): SymbolSelection => ({
  manualSymbols: [],
  ranking: {
    market_cap: null,
    volume: null,
    top_gainers: null,
    top_losers: null
  },
  excludedSymbols: [],
  excludedReasons: {},
  respectDefaultExclusions: true,
  leverageMode: 'uniform',
  leverageOverrides: {},
  positionOverrides: {},
  featureOverrides: {},
  maxListingAgeDays: null,
  rankingSort: 'alphabet',
  autoFillRecheck: false,
  excludeTopGainers: null,
  excludeTopLosers: null,
  excludeBottomVolume: null,
  excludeBottomMarketCap: null
});

export const createHedgeActivationSettings = (): HedgeActivationSettings => ({
  enabled: false,
  directions: ['long', 'short'],
  currentProfitRate: createProfitCondition(),
  indicators: createIndicatorConditions()
});

export const createDefaultAutoTradingSettings = (): AutoTradingSettings => ({
  logicName: '',
  leverage: 5,
  symbolCount: 1,
  assetMode: 'single',
  positionMode: 'one_way',
  capital: createCapitalSettings(),
  timeframe: '3m',
  symbolSelection: createSymbolSelection(),
  entry: createEntryConditions(),
  scaleIn: createScaleInConditions(),
  exit: createExitConditions(),
  stopLoss: createStopLossConditions(),
  hedgeActivation: createHedgeActivationSettings(),
  metadata: {
    lastSavedAt: null,
    lastValidatedAt: null
  }
});
