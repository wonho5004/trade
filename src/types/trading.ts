export type OrderSide = 'buy' | 'sell';

export interface ExchangeCredentials {
  apiKey: string;
  secret: string;
  password?: string;
}

export interface OrderRequest {
  symbol: string;
  side: OrderSide;
  quantity: number;
  price?: number;
  type: 'market' | 'limit';
}
