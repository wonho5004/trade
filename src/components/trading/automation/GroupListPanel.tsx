'use client';

import { useMemo, useState } from 'react';

import type { ConditionGroupNode, ConditionNode, IndicatorConditions, IndicatorLeafNode, StatusLeafNode } from '@/types/trading/auto-trading';
import { createIndicatorGroup, normalizeConditionTree } from '@/lib/trading/autoTradingDefaults';
import { collectGroupNodes, collectIndicatorNodes, ensureGroup, insertChild, removeNode } from '@/lib/trading/conditionsTree';
import { IndicatorEditModal } from './IndicatorEditModal';
import { normalizeSymbol as normalizePreviewSymbol } from '@/lib/trading/symbols';
import { useSymbolValidation } from '@/hooks/useSymbolValidation';
import { PreviewLauncher } from './PreviewLauncher';
import { GroupEditModal } from './GroupEditModal';

export function GroupListPanel({
  value,
  onChange,
  preview,
  groupPreviewInModal = true
}: {
  value: IndicatorConditions;
  onChange: (next: IndicatorConditions) => void;
  preview?: { symbol: string; symbolInput?: string; onSymbolChange?: (v: string) => void; quote?: 'USDT' | 'USDC'; datalistOptions?: string[]; interval: any; direction: 'long' | 'short'; indicatorSignals?: Record<string, boolean>; assumeSignalsOn?: boolean; onToggle?: (v: boolean) => void; onQuoteChange?: (q: 'USDT' | 'USDC') => void };
  groupPreviewInModal?: boolean;
}) {
  const [indicatorModal, setIndicatorModal] = useState<{ open: boolean; id: string | null }>({ open: false, id: null });
  const [groupModal, setGroupModal] = useState<{ open: boolean; id: string | null }>({ open: false, id: null });
  const [previewEnabled, setPreviewEnabled] = useState(false);

  const root = value.root as unknown as ConditionGroupNode;
  const baseGroups = useMemo(
    () => (root.kind === 'group' ? (root.children.filter((c) => c.kind === 'group') as ConditionGroupNode[]) : []),
    [root]
  );
  const directIndicators = useMemo(
    () => (root.kind === 'group' ? root.children.filter((c) => c.kind === 'indicator') : []),
    [root]
  );
  const isFallbackGroup = baseGroups.length === 0 && directIndicators.length > 0;
  const groups = isFallbackGroup ? [root] : baseGroups;
  const symValid = useSymbolValidation(preview ? (preview.symbolInput || preview.symbol) : '', (preview?.quote ?? 'USDT') as any);

  const addGroup = () => {
    const nextRoot: ConditionGroupNode = {
      ...root,
      operator: 'or',
      children: [...root.children, createIndicatorGroup('and', [])]
    };
    onChange({ root: normalizeConditionTree(nextRoot) } as any);
  };

  const removeGroup = (groupId: string) => {
    const next = ensureGroup(removeNode(root as unknown as ConditionNode, groupId) ?? (root as unknown as ConditionNode));
    onChange({ root: normalizeConditionTree(next) } as any);
  };

  const removeIndicator = (id: string) => {
    const next = ensureGroup(removeNode(root as unknown as ConditionNode, id) ?? (root as unknown as ConditionNode));
    onChange({ root: normalizeConditionTree(next) } as any);
  };

  const fmt = (n: any) => {
    const v = Number(n);
    if (!Number.isFinite(v)) return String(n);
    const s = v.toFixed(4);
    return s.replace(/0+$/, '').replace(/\.$/, '');
  };
  const labelOf = (type: string) => ({ bollinger: '볼린저 밴드', ma: '이동평균선', rsi: 'RSI', dmi: 'DMI/ADX', macd: 'MACD' } as Record<string, string>)[type] ?? type;
  const indicatorLabelMap = useMemo(() => {
    const map: Record<string, string> = {};
    try {
      collectIndicatorNodes(root as unknown as ConditionNode).forEach((n) => {
        map[n.id] = labelOf((n as any).indicator?.type ?? '');
      });
    } catch {}
    return map;
  }, [root]);

  const summary = (node: IndicatorLeafNode) => {
    const type = node.indicator.type;
    const config: any = node.indicator.config;
    const parts: string[] = [];
    if (type === 'ma') {
      if (config.period != null) parts.push(`period ${config.period}`);
      if (Array.isArray(config.actions) && config.actions.length > 0) parts.push(config.actions.join(','));
    } else if (type === 'bollinger') {
      if (config.length != null) parts.push(`len ${config.length}`);
      if (config.standardDeviation != null) parts.push(`${config.standardDeviation}σ`);
      if (config.band) parts.push(config.band);
    } else if (type === 'rsi') {
      if (config.period != null) parts.push(`period ${config.period}`);
      if (config.smoothing) parts.push(String(config.smoothing).toUpperCase());
      if (typeof config.threshold === 'number') parts.push(`th ${config.threshold}`);
      if (Array.isArray(config.actions) && config.actions.length > 0) parts.push(config.actions.join(','));
    } else if (type === 'dmi') {
      if (config.diPeriod != null) parts.push(`DI ${config.diPeriod}`);
      if (config.adxPeriod != null) parts.push(`ADX ${config.adxPeriod}`);
      if (config.diComparison) parts.push(config.diComparison);
      if (config.adxVsDi) parts.push(config.adxVsDi);
    } else if (type === 'macd') {
      const fast = config.fast ?? '-';
      const slow = config.slow ?? '-';
      const signal = config.signal ?? '-';
      parts.push(`(${fast},${slow},${signal})`);
      if (config.method) parts.push(config.method);
      if (config.comparison) parts.push(config.comparison);
      if (config.histogramAction) parts.push(`hist ${config.histogramAction}`);
    }
    const cmp = node.comparison as any;
    if (cmp?.kind === 'candle') parts.push(`cmp: candle ${String(cmp.field).toUpperCase()}(${cmp.reference}) ${cmp.comparator}`);
    if (cmp?.kind === 'value') parts.push(`cmp: ${cmp.comparator} ${fmt(cmp.value)}`);
    if (cmp?.kind === 'indicator') {
      const m = cmp.metric ? ` ${cmp.metric}` : '';
      const r = cmp.reference ? `(${cmp.reference})` : '';
      const tLabel = indicatorLabelMap[cmp.targetIndicatorId] ?? '다른 지표';
      parts.push(`cmp: ${tLabel}${m}${r} ${cmp.comparator}`.trim());
    }
    return parts.join(' · ');
  };

  const collectStatusNodes = (node: ConditionNode, acc: StatusLeafNode[] = []): StatusLeafNode[] => {
    if ((node as any).kind === 'status') {
      acc.push(node as any);
      return acc;
    }
    if ((node as any).kind === 'group') {
      (node as any).children.forEach((c: any) => collectStatusNodes(c, acc));
    }
    return acc;
  };

  const statusSummary = (s: StatusLeafNode) => {
    const label = s.metric === 'profitRate' ? '수익률' : s.metric === 'margin' ? '마진' : s.metric === 'buyCount' ? '매수횟수' : '진입경과';
    const unit = s.unit ? ` ${s.unit}` : '';
    return `${label} ${s.comparator} ${s.value}${unit}`;
  };

  return (
    <div className="space-y-3">
      <div className="flex items-center justify-between">
        <div className="flex items-center gap-2 text-[11px]">
          <span className="rounded border border-zinc-700 px-2 py-0.5 text-zinc-300">조건 그룹</span>
          <span className="rounded border border-zinc-700 px-2 py-0.5 text-zinc-400">그룹 간 OR(고정)</span>
          {preview ? (
            <PreviewLauncher
              conditions={value}
              symbol={preview.symbol}
              interval={preview.interval}
              direction={preview.direction}
              indicatorSignals={undefined}
            />
          ) : null}
          <button type="button" onClick={addGroup} className="rounded border border-emerald-500/60 px-3 py-1.5 text-xs font-semibold text-emerald-200">
            조건 그룹 추가
          </button>
        </div>
        <div className="flex items-center gap-2" />
      </div>

      {groups.length === 0 ? (
        <div className="text-[11px] text-zinc-500">그룹이 없습니다. ‘조건 그룹 추가’를 눌러 새 그룹을 만드세요.</div>
      ) : null}

      <div className="space-y-3">
        {groups.map((g, idx) => {
          const indicators = collectIndicatorNodes(g);
          const disableRemove = isFallbackGroup && g.id === root.id;
          return (
            <div key={g.id} className="space-y-2 rounded border border-zinc-800 bg-zinc-950/60 p-3">
              <div className="flex items-center justify-between">
                <div className="flex items-center gap-2 text-[11px]">
                  <span className="rounded border border-zinc-700 px-2 py-0.5 text-zinc-300">조건 그룹 {idx + 1}</span>
                  <span className="rounded border border-zinc-700 px-2 py-0.5 text-zinc-400">그룹 내 AND(고정)</span>
                </div>
                <div className="flex items-center gap-2">
                  <button
                    type="button"
                    onClick={() => setGroupModal({ open: true, id: g.id })}
                    className="rounded border border-zinc-700 px-2 py-0.5 text-[11px] text-zinc-300"
                  >
                    그룹 편집
                  </button>
                  <button
                    type="button"
                    onClick={() => removeGroup(g.id)}
                    disabled={disableRemove}
                    className="rounded border border-zinc-700 px-2 py-0.5 text-[11px] text-rose-300"
                  >
                    그룹 삭제
                  </button>
                </div>
              </div>

              {(() => {
                const statuses = collectStatusNodes(g as any);
                const hasAny = indicators.length > 0 || statuses.length > 0;
                if (!hasAny) return <div className="text-[11px] text-zinc-500">지표/상태 조건이 없습니다. ‘그룹 편집’에서 추가하세요.</div>;
                return (
                  <ul className="space-y-1">
                    {indicators.map((n) => (
                      <li key={n.id} className="flex items-center justify-between gap-2 rounded border border-zinc-800 bg-zinc-950/40 p-2">
                        <div className="min-w-0">
                          <p className="truncate text-[11px] text-zinc-400">{summary(n)}</p>
                        </div>
                        <div className="flex items-center gap-2">
                          <button
                            type="button"
                            onClick={() => setIndicatorModal({ open: true, id: n.id })}
                            className="rounded border border-zinc-700 px-2 py-0.5 text-[11px] text-zinc-300"
                          >
                            지표 편집
                          </button>
                          <button
                            type="button"
                            onClick={() => removeIndicator(n.id)}
                            className="rounded border border-zinc-700 px-2 py-0.5 text-[11px] text-rose-300"
                          >
                            삭제
                          </button>
                        </div>
                      </li>
                    ))}
                    {statuses.map((s) => (
                      <li key={s.id} className="flex items-center justify-between gap-2 rounded border border-zinc-800 bg-zinc-950/40 p-2">
                        <div className="min-w-0">
                          <p className="truncate text-[11px] text-zinc-400">상태 · {statusSummary(s)}</p>
                        </div>
                      </li>
                    ))}
                  </ul>
                );
              })()}
            </div>
          );
        })}
      </div>

      <IndicatorEditModal
        open={indicatorModal.open}
        title="지표 편집"
        conditions={value}
        indicatorId={indicatorModal.id ?? ''}
        onChange={onChange}
        onClose={() => setIndicatorModal({ open: false, id: null })}
      />
      <GroupEditModal
        open={groupModal.open}
        title="그룹 편집"
        conditions={value}
        groupId={groupModal.id ?? ''}
        onChange={onChange}
        onClose={() => setGroupModal({ open: false, id: null })}
        preview={groupPreviewInModal && preview ? {
          symbol: preview.symbol,
          symbolInput: preview.symbolInput,
          onSymbolChange: preview.onSymbolChange,
          quote: preview.quote,
          interval: preview.interval,
          direction: preview.direction,
          indicatorSignals: preview.indicatorSignals,
          assumeSignalsOn: preview.assumeSignalsOn,
          onToggle: preview.onToggle
        } : undefined}
      />
    </div>
  );
}
