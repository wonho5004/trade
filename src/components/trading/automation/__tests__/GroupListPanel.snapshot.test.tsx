// @ts-nocheck
import React from 'react';
import { render } from '@testing-library/react';

import { GroupListPanel } from '../GroupListPanel';
import { createIndicatorEntry, createIndicatorLeaf, createIndicatorGroup } from '@/lib/trading/autoTradingDefaults';

function sampleConditions() {
  const rsi = createIndicatorEntry('rsi');
  (rsi.config as any).period = 14;
  const leaf = createIndicatorLeaf(rsi, { kind: 'none' } as any);
  const group = createIndicatorGroup('and', [leaf]);
  return { root: createIndicatorGroup('or', [group]) } as any;
}

describe('GroupListPanel snapshots', () => {
  it('renders group summary consistently', () => {
    const value = sampleConditions();
    const { container } = render(<GroupListPanel value={value} onChange={() => {}} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

