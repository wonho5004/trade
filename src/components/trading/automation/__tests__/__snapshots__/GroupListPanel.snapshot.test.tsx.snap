// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupListPanel snapshots renders group summary consistently 1`] = `
<div
  class="space-y-3"
>
  <div
    class="flex items-center justify-between"
  >
    <div
      class="flex items-center gap-2 text-[11px]"
    >
      <span
        class="rounded border border-zinc-700 px-2 py-0.5 text-zinc-300"
      >
        조건 그룹
      </span>
      <span
        class="rounded border border-zinc-700 px-2 py-0.5 text-zinc-400"
      >
        그룹 간 OR(고정)
      </span>
    </div>
    <button
      class="rounded border border-emerald-500/60 px-3 py-1.5 text-xs font-semibold text-emerald-200"
      type="button"
    >
      조건 그룹 추가
    </button>
  </div>
  <div
    class="space-y-3"
  >
    <div
      class="space-y-2 rounded border border-zinc-800 bg-zinc-950/60 p-3"
    >
      <div
        class="flex items-center justify-between"
      >
        <div
          class="flex items-center gap-2 text-[11px]"
        >
          <span
            class="rounded border border-zinc-700 px-2 py-0.5 text-zinc-300"
          >
            조건 그룹 
            1
          </span>
          <span
            class="rounded border border-zinc-700 px-2 py-0.5 text-zinc-400"
          >
            그룹 내 AND(고정)
          </span>
        </div>
        <div
          class="flex items-center gap-2"
        >
          <button
            class="rounded border border-zinc-700 px-2 py-0.5 text-[11px] text-zinc-300"
            type="button"
          >
            그룹 편집
          </button>
          <button
            class="rounded border border-zinc-700 px-2 py-0.5 text-[11px] text-rose-300"
            type="button"
          >
            그룹 삭제
          </button>
        </div>
      </div>
      <ul
        class="space-y-1"
      >
        <li
          class="flex items-center justify-between gap-2 rounded border border-zinc-800 bg-zinc-950/40 p-2"
        >
          <div
            class="min-w-0"
          >
            <p
              class="truncate text-[11px] text-zinc-400"
            >
              period 14 · SMA · th 50
            </p>
          </div>
          <div
            class="flex items-center gap-2"
          >
            <button
              class="rounded border border-zinc-700 px-2 py-0.5 text-[11px] text-zinc-300"
              type="button"
            >
              지표 편집
            </button>
            <button
              class="rounded border border-zinc-700 px-2 py-0.5 text-[11px] text-rose-300"
              type="button"
            >
              삭제
            </button>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
